appnav/src/main/kotlin/io/element/android/appnav/di/DefaultSyncOrchestrator.kt

```diff
@@ -0,0 +1,148 @@
+    override fun start() {
+        if (!started.compareAndSet(false, true)) {
+            Timber.tag(tag).d("already started, exiting early")
+            return
+        }
+
+        Timber.tag(tag).d("start observing the app and network state")
+
+        if (syncService.syncState.value != SyncState.Running) {
+            Timber.tag(tag).d("initial startSync")
+            baseCoroutineScope.launch(dispatchers.io) {
+                try {
+                    initialSyncMutex.lock()
+                    syncService.startSync()
+
+                    // Wait until it's running
+                    syncService.syncState.first { it == SyncState.Running }
+                } finally {
+                    initialSyncMutex.unlock()
+                }
+            }
+        }
+
+        coroutineScope = CoroutineScope(baseCoroutineScope.coroutineContext + CoroutineName(tag) + dispatchers.io)
+
+        coroutineScope?.launch {
+            // Wait until the initial sync is done, either successfully or failing
+            initialSyncMutex.lock()
+
+            combine(
+                // small debounce to avoid spamming startSync when the state is changing quickly in case of error.
+                syncService.syncState.debounce(100.milliseconds),
+                networkMonitor.connectivity,
+                appForegroundStateService.isInForeground,
+                appForegroundStateService.isInCall,
+                appForegroundStateService.isSyncingNotificationEvent,
+            ) { syncState, networkState, isInForeground, isInCall, isSyncingNotificationEvent ->
+                val isAppActive = isInForeground || isInCall || isSyncingNotificationEvent
+                val isNetworkAvailable = networkState == NetworkStatus.Connected
+
+                Timber.tag(tag).d("isAppActive=$isAppActive, isNetworkAvailable=$isNetworkAvailable")
+                if (syncState == SyncState.Running && !isAppActive) {
+                    // Don't stop the sync immediately, wait a bit to avoid starting/stopping the sync too often
+                    delay(3.seconds)
+                    SyncStateAction.StopSync
+                } else if (syncState != SyncState.Running && isAppActive && isNetworkAvailable) {
+                    SyncStateAction.StartSync
+                } else {
+                    SyncStateAction.NoOp
+                }
+            }
+                .distinctUntilChanged()
+                .collect { action ->
+                    when (action) {
+                        SyncStateAction.StartSync -> {
+                            syncService.startSync()
+                        }
+                        SyncStateAction.StopSync -> {
+                            syncService.stopSync()
+                        }
+                        SyncStateAction.NoOp -> Unit
+                    }
+                }
+        }
+    }
+
+    /**
+     * Stop observing the app state and network state.
+     */
+    override fun stop() {
```

Comment:
Since this method is only used for test, you should remove it.
The background coroutine scope can be used from the TestScope, and can be provided in the constructor parameter.